### PR TITLE
AngularMigrate: Auto migrate graph to multiple panels

### LINKED ISSUE #83979: Dashboard: Support migrations from one to multiple panel types
as our angular removal approaches, we need to migrate old panels to new panels. one of these panels is `Graph (old)` whose purpose as TimeSeries, Histogram, and BarChart has been split into multiple panels.

currently @grafana/dataviz-squad does not see a way to migrate `Graph (old)` to each of these panels based on the old `xaxis.mode` option, which is what determined the behavior of the old panel.

plz halp!

### PR DESCRIPTION
**Pull request by torkelo**
Fixes #83979

* [x] Refactor and centralize autoMigrateFrom logic inside a function in PanelModel 
* [x] Simplify overly complex feature toggle checking code 
* [x] Add support to migrate graph to timeseries, barchart and histogram depending on panel.xaxis.mode option 

graph => barchart 
* [x] Add foundation for graph => barchart migration
* [x] Add reduce transform
* [x] Unit tests 

graph => histogram
* [x] Basic migration
* [x] Unit test

### CODE REVIEW

File: public/app/plugins/panel/histogram/migrations.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  prevPluginId,
+  prevOptions,
+  prevFieldConfig
+) => {
+  if (prevPluginId === 'graph') {
+    const graphOptions: GraphOptions = prevOptions.angular;
+
+    if (graphOptions.xaxis?.mode === 'histogram') {
+      return {
+        combine: true,
```
Comment (leeoniya) [inline public/app/plugins/panel/histogram/migrations.ts:17]: i don't think we want this. unless the old panel combined multiple series into one. iiuc the old one simple has a reducer per series and can optionally stack the result of that.

File: public/app/plugins/panel/barchart/migrations.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { PanelTypeChangedHandler } from '@grafana/data';
+
+/*
+ * This is called when the panel changes from another panel
+ */
+export const changeToBarChartPanelMigrationHandler: PanelTypeChangedHandler = (
```
[the remark this answers is not in the document]
Comment (nmarrs) [reply public/app/plugins/panel/barchart/migrations.ts:6]: [done](https://github.com/grafana/grafana/pull/83992/commits/24bd3a95f0930cc44cc1b6864de3f641222466d8) 


Recording TBD

File: public/app/features/dashboard/state/PanelModel.ts [hunk 3]
```diff
@@ -729,3 +729,47 @@ export function stringifyPanelModel(panel: PanelModel) {
...
+
+  if (panel.type === 'grafana-piechart-panel' && config.featureToggles.autoMigratePiechartPanel) {
+    return 'piechart';
+  }
+
+  if (panel.type === 'grafana-worldmap-panel' && config.featureToggles.autoMigrateWorldmapPanel) {
+    return 'geomap';
+  }
+
+  if (panel.type === 'singlestat' || config.featureToggles.autoMigrateStatPanel) {
```
[the remark this answers is not in the document]
Comment (nmarrs) [reply public/app/features/dashboard/state/PanelModel.ts]: Also in its current state the logic will migrate all panels to a stat panel if the autoMigrateStatPanel is set to true 😅 I think we just need to modify this to an `&&` to fix this


Recording TBD

Addressed the above [with this commit](https://github.com/grafana/grafana/pull/83992/commits/0f8e147d26bfa199a6e062e7c9b2ebc32410bd67)

**Review by nmarrs — APPROVED** [review]
Tested locally and fixed a minor logic issue with stat feature toggle. This approach is very clean! 🚀

File: public/app/plugins/panel/histogram/migrations.ts (hunk 1 above)
Comment (leeoniya) [reply public/app/plugins/panel/histogram/migrations.ts:17]: actually nvm, i think my brain is in `series` / BarChart mode. but still, this should only enabled when stacking is off in the old panel?

File: public/app/features/dashboard/state/PanelModel.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryantxu) [reply public/app/features/dashboard/state/PanelModel.ts]: > grafana-singlestat-panel

maybe only if angular is actually disabled -- if they have `grafana-singlestat-panel` installed, it was an intentional move to not migrate

[the remark this answers is not in the document]
Comment (nmarrs) [reply public/app/features/dashboard/state/PanelModel.ts]: @ryantxu even if intent was not to migrate if we are forcing migration anyways why not include that variant to be controlled by the `autoMigrateStatPanel` feature toggle?

**Pull request merged**